### PR TITLE
[MPS] Store philox counter as part of the RNG state

### LINKED ISSUE #130613: `torch.Generator.set_state()` does not correctly restore state on MPS.
### 🐛 Describe the bug

Restoring the state of a `torch.Generator` does not correctly reproduce subsequent random number generations as the original `torch.Generator` on mps.

To reproduce:

```python
import torch
device = torch.device("mps")

gen = torch.Generator(device=device)
gen = gen.manual_seed(42)
probs = torch.ones(3, 3).to(device)
a = torch.empty_like(probs).exponential_(1, generator=gen)
b = torch.empty_like(probs).exponential_(1, generator=gen)

state = gen.get_state()
c = torch.empty_like(probs).exponential_(1, generator=gen)

gen2 = torch.Generator(device=device)
gen2.set_state(state)
probs = torch.ones(3, 3).to(device)
d = torch.empty_like(probs).exponential_(1, generator=gen2)

a, b, c, d, (c == d).all()

outputs:
(tensor([[1.6993e+00, 8.8699e-01, 8.1656e-01],
         [1.4660e+00, 4.1455e-01, 6.0873e-01],
         [1.1242e+00, 4.4069e-04, 5.2553e-01]], device='mps:0'),
 tensor([[0.0684, 0.3959, 0.7246],
         [0.2561, 1.7915, 0.8894],
         [0.2813, 0.8345, 0.6247]], device='mps:0'),
 tensor([[0.8867, 0.2788, 0.8667],
         [1.1411, 0.8153, 0.3937],
         [0.5605, 0.2341, 0.0578]], device='mps:0'),
 tensor([[1.6993e+00, 8.8699e-01, 8.1656e-01],
         [1.4660e+00, 4.1455e-01, 6.0873e-01],
         [1.1242e+00, 4.4069e-04, 5.2553e-01]], device='mps:0'),
 tensor(False, device='mps:0'))
```

and the same code on CUDA:

```python
import torch
device = torch.device("cuda")

gen = torch.Generator(device=device)
gen = gen.manual_seed(42)
probs = torch.ones(3, 3).to(device)
a = torch.empty_like(probs).exponential_(1, generator=gen)
b = torch.empty_like(probs).exponential_(1, generator=gen)

state = gen.get_state()
c = torch.empty_like(probs).exponential_(1, generator=gen)

gen2 = torch.Generator(device=device)
gen2.set_state(state)
probs = torch.ones(3, 3).to(device)
d = torch.empty_like(probs).exponential_(1, generator=gen2)

a, b, c, d, (c == d).all()

outputs:
(tensor([[0.4895, 4.5993, 0.9203],
         [3.2112, 1.8562, 0.7288],
         [0.3062, 0.9014, 0.6560]], device='cuda:0'),
 tensor([[0.0124, 2.0487, 0.5761],
         [0.6499, 0.2951, 0.5184],
         [0.0359, 0.1077, 0.2575]], device='cuda:0'),
 tensor([[0.1913, 1.5475, 0.3158],
         [1.1365, 0.2811, 1.5454],
         [0.0750, 1.9363, 1.0579]], device='cuda:0'),
 tensor([[0.1913, 1.5475, 0.3158],
         [1.1365, 0.2811, 1.5454],
         [0.0750, 1.9363, 1.0579]], device='cuda:0'),
 tensor(True, device='cuda:0'))
```

Any input would be greatly appreciated. I'd be happy to try contribute a fix if it's within my understanding.

### Versions

```bash
(venv) combuter:torchtune salmanmohammadi$ python collect_env.py
Collecting environment information...
PyTorch version: 2.5.0.dev20240707
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.4 (main, Jun  6 2024, 18:26:44) [Clang 15.0.0 (clang-1500.3.9.4)] (64-bit runtime)
Python platform: macOS-14.4.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.5.0.dev20240707
[pip3] torchao==0.3.0
[pip3] torchao==0.3.0
[pip3] torchaudio==2.4.0.dev20240707
[pip3] torchtune==0.0.0
[pip3] torchvision==0.20.0.dev20240707
```

cc @pbelevich @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

### PR DESCRIPTION
**Pull request by qqaatw**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130662

Fixes #130613

### CODE REVIEW

File: test/test_mps.py [hunk 1]
```diff
@@ -7781,22 +7781,27 @@ def test_mps_generator(self):
...
         # in this case, the random results must differ from the last generated random results
         self.assertNotEqual(mps_x, mps_y)
 
+        # mps_x was produced by g_state, we use it as our reference mps_y.
+        mps_y = mps_x
+
         # restore the previously saved state, and the results should match again
         g_mps.set_state(g_state)
         mps_x = torch.randn(5, device='mps', generator=g_mps)
         self.assertEqual(mps_x, mps_y)
```
Comment (qqaatw) [inline test/test_mps.py:7802]: The previous testing approach was flawed. It did not consider Philox's offset at all. The reason that the `assertEqual` passed was that `mps_y` was generated with `offset = 0` and the `g_state` should have had carry `offset = 1` but it didn't. As a result, here when we restored from `g_state`, we implicitly set `offset = 0`, and `mps_x` and `mps_y` coincidentally matched.

File: test/test_mps.py [hunk 2]
```diff
@@ -7781,22 +7781,27 @@ def test_mps_generator(self):
...
 
+    @serialTest()
```
Comment (qqaatw) [inline test/test_mps.py:7804]: Feels like this should be serial as random functions in other tests may cause the Philox's counter to increase inadvertently.

File: test/test_mps.py [hunk 3]
```diff
@@ -7806,19 +7811,25 @@ def test_default_mps_generator(self):
...
+        # generate random numbers with offset `1`
         mps_x = torch.randn(5, device='mps')
         # in this case, the random results must differ from the last generated random results
         self.assertNotEqual(mps_x, mps_y)
+        # since we called randn twice after seeding, the offset should be 2
+        self.assertEqual(torch.mps._get_default_mps_generator().get_offset(), 2)
 
-        # restore the previously saved state, and the results should match again
+        # mps_x was produced by g_state, we use it as our reference mps_y.
+        mps_y = mps_x
```
Comment (qqaatw) [inline test/test_mps.py:7830]: Explained in the above comment

**Comment by qqaatw** [comment]
@malfet @kulinseth Can you please take a look?

**Review by malfet — APPROVED** [review]
LGTM

File: aten/src/ATen/mps/MPSGeneratorImpl.mm [hunk 4]
```diff
@@ -65,32 +65,39 @@ Generator createMPSGenerator(uint64_t seed_val) {
 c10::intrusive_ptr<c10::TensorImpl> MPSGeneratorImpl::get_state() const {
   static const size_t states_size = mps::detail::PHILOX_STATE_N * sizeof(uint32_t);
   static const size_t seed_size = sizeof(uint64_t);
-  static const size_t total_size = states_size + seed_size;
+  static const size_t offset_size = sizeof(uint64_t);
```
Comment (malfet) [inline aten/src/ATen/mps/MPSGeneratorImpl.mm:68]: Not sure why it needs to be static (and why couldn't it be a constexpr)
```suggestion
  constexpr auto offset_size = sizeof(uint64_t);
```

File: aten/src/ATen/mps/MPSGeneratorImpl.mm [hunk 5]
```diff
@@ -65,32 +65,39 @@ Generator createMPSGenerator(uint64_t seed_val) {
...
+  static const size_t total_size = states_size + seed_size + offset_size;
 
   auto state_tensor = at::detail::empty_cpu(
       {(int64_t)total_size}, ScalarType::Byte, std::nullopt, std::nullopt, std::nullopt, std::nullopt);
   auto rng_state = state_tensor.data_ptr<uint8_t>();
   auto current_seed = this->current_seed();
+  auto current_offset = this->get_offset();
```
Comment (malfet) [inline aten/src/ATen/mps/MPSGeneratorImpl.mm:75]:
```suggestion
  auto current_offset = this->get_offset();
  static_assert(sizeof(decltype(current_offset) == offset_size, "current_offset size is wrong");
```

**Pull request closed without merging**